runtime/tests/checker/resources_test.go

```diff
@@ -3984,3 +4026,35 @@ func TestCheckInvalidOptionalResourceCoalescingRightSide(t *testing.T) {
 	assert.IsType(t, &sema.ResourceLossError{}, errs[1])
 	assert.IsType(t, &sema.ResourceLossError{}, errs[2])
 }
+
+// https://github.com/dapperlabs/flow-go/issues/3407
+//
+// Check that an function's return information
+// does not influence another function's return information.
+//
+func TestCheckInvalidResourceLossInNestedContractResource(t *testing.T) {
+
+	_, err := ParseAndCheckWithPanic(t, `
+
+      pub contract C {
+
+          resource R {
+
+              fun foo(r: @R) {
+                  if let r2 <- r as? @R {
+                      destroy r2
+                  }
+              }
+          }
+
+          pub fun bar() {
```

Comment:
So previously, this would have lead to `bar()` not passing checking due to a resource loss error, correct?

Reply:
No, actually it's the other way round: `bar` is checked first, and the `return` statement set the `Returns` flag. Then `foo` is checked and the `Returns` flag is already set to true, so the whole of `foo` looks like dead-code to the resource tracking code: the function has already returned, the function is therefore valid – which is incorrect, because the flag was not reset and there is actually a resource loss of `r` in `foo` if the conditional cast fails.